mods/stairs/init.lua

```diff
@@ -126,87 +129,35 @@ function stairs.register_slab(subname, recipeitem, groups, images, description,
-			local param2 = 0
-
-			local n0_is_upside_down = (n0.name == "stairs:slab_" .. subname and
-					n0.param2 >= 20)
-
-			if n0.name == "stairs:slab_" .. subname and not n0_is_upside_down and
-					p0.y + 1 == p1.y then
-				slabpos = p0
-				slabnode = n0
-			elseif n1.name == "stairs:slab_" .. subname then
-				slabpos = p1
-				slabnode = n1
-			end
-			if slabpos then
-				-- Remove the slab at slabpos
-				minetest.remove_node(slabpos)
-				-- Make a fake stack of a single item and try to place it
-				local fakestack = ItemStack(recipeitem)
-				fakestack:set_count(itemstack:get_count())
-
-				pointed_thing.above = slabpos
-				local success
-				fakestack, success = minetest.item_place(fakestack, placer,
-					pointed_thing)
-				-- If the item was taken from the fake stack, decrement original
-				if success then
-					itemstack:set_count(fakestack:get_count())
-				-- Else put old node back
-				else
-					minetest.set_node(slabpos, slabnode)
+		on_rightclick = function(pos, node, clicker, itemstack, pointed_thing)
+			local under = minetest.get_node(pointed_thing.under)
+			local dir = minetest.dir_to_facedir(vector.subtract(
+				pointed_thing.above, pointed_thing.under), true)
+
+			local wield_item = clicker:get_wielded_item():get_name()
+			local player_name = clicker:get_player_name()
+
+			-- combine two slabs
+			if wield_item == node.name and
+					slab_trans_dir[math.floor(under.param2 / 4)] == dir then
+				if minetest.is_protected(pos, player_name) and not
+						minetest.check_player_privs(placer, "protection_bypass") then
+					minetest.record_protection_violation(pos, player_name)
+					return
 				end
-				return itemstack
-			end
-			
-			-- Upside down slabs
-			if p0.y - 1 == p1.y then
-				-- Turn into full block if pointing at a existing slab
-				if n0_is_upside_down  then
-					-- Remove the slab at the position of the slab
-					minetest.remove_node(p0)
-					-- Make a fake stack of a single item and try to place it
-					local fakestack = ItemStack(recipeitem)
-					fakestack:set_count(itemstack:get_count())
-
-					pointed_thing.above = p0
-					local success
-					fakestack, success = minetest.item_place(fakestack, placer,
-						pointed_thing)
-					-- If the item was taken from the fake stack, decrement original
-					if success then
-						itemstack:set_count(fakestack:get_count())
-					-- Else put old node back
-					else
-						minetest.set_node(p0, n0)
-					end
-					return itemstack
+				minetest.set_node(pos, {name = recipeitem, param2 = node.param2})
+				if not minetest.setting_getbool("creative_mode") then
+					itemstack:take_item()
 				end
-
-				-- Place upside down slab
-				param2 = 20
+				return itemstack
 			end
 
-			-- If pointing at the side of a upside down slab
-			if n0_is_upside_down and p0.y + 1 ~= p1.y then
-				param2 = 20
+			-- else attempt to place node with proper param2
+			minetest.item_place_node(ItemStack(wield_item), clicker, pointed_thing, node.param2)
```

Comment:
You can place upside down slabs without issues. If you place any slab against the side of an already-upside down slab, it will also be upside-down.

You can also combine an upside down slab from below with another slab.